compiler/qsc_eval/src/val.rs

```diff
@@ -23,11 +23,65 @@ pub enum Value {
     Pauli(Pauli),
     Qubit(Qubit),
     Range(Option<i64>, i64, Option<i64>),
-    Result(bool),
+    Result(Result),
     String(Rc<str>),
     Tuple(Rc<[Value]>),
 }
 
+#[derive(Clone, Copy, Debug)]
+pub enum Result {
+    Val(bool),
+    Id(usize),
+}
+
+impl From<Result> for bool {
+    fn from(val: Result) -> Self {
+        match val {
+            Result::Val(v) => v,
+            Result::Id(_) => panic!("cannot convert Result::Id to bool"),
+        }
+    }
+}
+
+impl From<&Result> for bool {
+    fn from(val: &Result) -> Self {
+        match val {
+            Result::Val(v) => *v,
+            Result::Id(_) => panic!("cannot convert Result::Id to bool"),
+        }
+    }
+}
+
+impl From<bool> for Result {
+    fn from(val: bool) -> Self {
+        Self::Val(val)
+    }
+}
+
+impl From<usize> for Result {
+    fn from(val: usize) -> Self {
+        Self::Id(val)
+    }
+}
+
+impl From<Result> for usize {
+    fn from(val: Result) -> Self {
+        match val {
+            Result::Val(..) => panic!("cannot convert Result::Val to usize"),
+            Result::Id(v) => v,
+        }
+    }
+}
+
+impl PartialEq for Result {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (Result::Val(a), Result::Val(b)) => a == b,
+            (Result::Id(_), _) | (_, Result::Id(_)) => panic!("cannot compare Result::Id"),
```

Comment:
Panicking in PartialEq also seems iffy.

Reply:
Same thinking here, it is based on the things that shouldn't occur because the pass would disallow them.

Reply:
I can go back to what I had before, which is to use the `derive` of `PartialEq`. The behavior there doesn't match the goal, which is to panic on disallowed operations, but that behavior is only temporary until the analysis pass gets checked in, so no reason to over index on that. That does mean that if there is a comparison between a `Result::Id` and the literals for `One` or `Zero` it will equal neither of them, ie `res == One` and `res == Zero` would both return false. Again, this is only a concern for the window in which this code gets checked in before #494, so not really a big worry.

Reply:
The PartialEq impl for Value already allows you to make comparisons forbidden by Q#, like comparing global callables or closures with other values, or comparing values of different Q# types. Do you think Result comparisons are different from those?

Reply:
Nope, it's the same. Again, I was trying to get specific behavior as a temporary solution until #494 merges. This is moot once that guard is in place, and then they are exactly the same as global callables in that regard.